Twister: ADC-based power measurement harness

hakehuang opened the pull request:
1. Create adc_power_shield sample for any board with adc can be used for power measurement.
2. Add general_adc_platform power shiled, which can be used for power measurement with the sample application.
3. the power data is stored in the power_shield folder for further analysis
4. add power measure config in testsuite.

hakehuang committed:
samples: adc: add power measure application

add power measure sample for general_adc_platform
power shiled, to record power/voltage/current

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
tests: scripts: add general power shield support

1. general power shield work with the adc_power_shield sample
2. config samples to config probe
3. add readme
4. enable parse shell command outputs, see readme for details

Signed-off-by: Hake Huang <hake.huang@nxp.com>

remove document

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
samples: boards: nxp: enable power shield on mimxrt595_evk

enable power shield testing on mimxrt595_evk

```
scripts/twister --device-testing --device-serial /dev/ttyACM0
-T samples/boards/nxp/mimxrt595_evk/system_off
-X pm_probe:/dev/ttyACM1,115200
```

an power_shield folder will be created in the build path
whith measured voltage/current/power

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang commented on samples/drivers/adc/adc_power_shield/sample.yaml:
this sample is not to sample vendor adc usage, but to enable user as many options boards to use as power shield.

hakehuang replied:
thanks @JarmouniA PR #94585 looks good, I will update according after your PR merged

hakehuang commented:
will move to active after #94585 get merged

hakehuang commented on samples/drivers/adc/adc_power_shield/boards/s32z2xxdc2_s32z270_rtu0.overlay:
```diff
@@ -0,0 +1,55 @@
+	aliases {
+		adc0 = &sar_adc0;
+	};
+	zephyr,user {
```
> newline line 13, fix all these issues in the whole PR

@nordicjm thanks for pointing out this, just wonder as I copied those file from adc_sequence sample, are there new rules enforced recently?

nordicjm replied:
yes, https://docs.zephyrproject.org/latest/contribute/style/devicetree.html was added this year

dsseng reviewed (commented):
Thanks, this is a really great addition

hakehuang commented on scripts/pylib/power-twister-harness/requirements-dev.txt:
```diff
@@ -0,0 +1,31 @@
+scipy>=1.7
+matplotlib>=3.3
+
+# Additional development tools
+ipython>=8.0.0
+jupyter>=1.0.0
```
right, it is added by my local setting, will remove it

hakehuang commented:
https://github.com/zephyrproject-rtos/zephyr/pull/95056 I define a dft (device for test) which can also extend to multi-application build

hakehuang commented on samples/drivers/adc/adc_power_shield/README.rst:
```diff
@@ -0,0 +1,80 @@
+.. zephyr:code-sample:: adc_power_shield
+   :name: Analog-to-Digital Converter (ADC) power shield sample
+   :relevant-api: adc_interface
+
+   Read analog inputs from ADC channels, using a sequence.
+
+Overview
+********
+
+This sample is to enable general power shield for platfroms that have ADC support.
+
+
+Building and Running
+********************
+
+Make sure that the ADC is enabled (``status = "okay";``) and has each channel as a
+child node, with your desired settings like gain, reference, or acquisition time and
+oversampling setting (if used). It is also needed to provide an alias ``adc0`` for the
+desired adc. See :zephyr_file:`boards/nrf52840dk_nrf52840.overlay
+<samples/drivers/adc/adc_dt/boards/nrf52840dk_nrf52840.overlay>` for an example of
+such setup.
```
@kartben, the sample application is reused from adc_sequence sample. The idea here is to create a general power shield, which means any adc featured zephyr platform can be used as a power shield. so we can take any of those 100s boards and flashed with this sample, you can use them by power_shield fixture for measuring power, see the https://github.com/zephyrproject-rtos/zephyr/blob/af6248f8b7509f505982b8fe5d6e9a6efd8bc0f5/scripts/pylib/power-twister-harness/README.rst this is a full picture for the usage. 
the sample readme is just describe how the sample runs.

hakehuang replied:
<img width="828" height="446" alt="image" src="https://github.com/user-attachments/assets/625a4423-14f6-45e8-b6ef-d336d8a588bf" />
the is a demo setup for general power measurement.

hakehuang commented:
@nashif please take a look.

hakehuang committed:
tests: pm: power_mgmt_soc: enable measurement on rt1060

enable power measure on power_mgmt_soc for
mimxrt1060_evk@C

Signed-off-by: Hake Huang <hake.huang@nxp.com>

bjarki-andreasen reviewed (changes requested):
Has it been considered to just use the ADC shell and a "blank" application? Setting up and performing a reading can be done directly using the shell, no custom parsing of console input/output should be needed.

hakehuang commented:
> Has it been considered to just use the ADC shell and a "blank" application? Setting up and performing a reading can be done directly using the shell, no custom parsing of console input/output should be needed.

the main purpose of this harness is for CI. As we want to a unified test approach, if in a blank application, you will have to customize it by you own, in this case, you are not using CI.

hakehuang commented:
> Has it been considered to just use the ADC shell and a "blank" application? Setting up and performing a reading can be done directly using the shell, no custom parsing of console input/output should be needed.

The main purpose of this harness is for CI. As we want to a unified test approach, if in a blank application, you will have to customize it by your own, in this case, you are not using CI.

bjarki-andreasen commented:
> > We already have a twister harness for the shell, can it not be reused? and regarding the specific ADC channel, that can be added in an overlay when building surely, like this PR is already doing? We can also add a custom shell command to use specifically if needed, no need to implement command line parsing in the sample
> 
> @bjarki-andreasen  if you look at my PR, I add a 'dft' which reuse the whole pytest serial handler see https://github.com/zephyrproject-rtos/zephyr/blob/815f2b6cc0daa0f0cd10c582cb6334388fad13d8/scripts/pylib/power-twister-harness/conftest.py#L130. ADC channel is already controlled by overlay, but the problem is that we need map the channel to the target power rails see below definition:
> the channels_p and channel_n to the power rail "VDD_CORE" / "VDD_IO" is by your board setting definition, this could be different pre your test harness setting, and this has to be customized, here is use a yaml to control it.
> ```
> device_id: "adc_power_monitor_01"
> routes:
>   - id: 0
>     name: "VDD_CORE"
>     shunt_resistor: 0.1  # ohms
>     gain: 1.0
>     type: "single"
>     channels:
>       channels_p: 4
>       channels_n: 0
>   - id: 1
>     name: "VDD_IO"
>     shunt_resistor: 0.1  # ohms
>     gain: 1.0
>     type: "single"
>     channels:
>       channels_p: 7
>       channels_n: 3
> calibration:
>   offset: 0.0
>   scale: 1.0
> ```
> so nowhat where to add, we either add shell command or use a dedicated power shield sample application, we have to parse the shell/command output to get the board settings on its probe capacities, see below.
> 
> ```
> {
>   "channel_count": 2,
>   "resolution": 12,
>   "channels": {
>     "0": {
>       "mode": "single",
>       "verf_mv": 3300
>     },
>     "3": {
>       "mode": "single",
>       "verf_mv": 3300
>     },
>     "4": {
>       "mode": "single",
>       "verf_mv": 3300
>     },
>     "7": {
>       "mode": "single",
>       "verf_mv": 3300
>     }
>   }
> }
> ``` 
> and we need map the probe capacities to the target measure power rails as I mentioned above. My understanding is that you propose to use a series of shell command to replace the power measure sample which will be running on power shield. this will add adc shell linked with power calculations I am not sure this will accepted by ADC shell owner? As in this PR, we have an independent sample to host everything, move those to shell command needs approval, can you help to confirm on my understanding and if accepted by ADC shell group I can move the implement to shell.
> 

No changes to the adc shell is needed. We would just create a set of custom shell commands as part of the power shield sample. The issue is not with having a sample at all, we need that, but with how the serial communication is implemented. We can freely create sample specific shell commands, so why not use that?

hakehuang committed:
testsuite: option: add power measure option to test

add power measure options to test

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
tests: pm: power_residency_time: add power measure check

demo the power measure with frdm_rw612 boards
keep the system in final power state for measurement

Signed-off-by: Hake Huang <hake.huang@nxp.com>

bjarki-andreasen reviewed (dismissed):
I'm surprised how much python code is needed to make this work

hakehuang commented:
@kartben I update the sample application readme and reduce them to platforms that I tested. Please take a look. Thanks,

JarmouniA commented on samples/drivers/adc/adc_power_measure/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(ADC)
```
```suggestion
project(adc_voltage_measure)
```

JarmouniA commented on scripts/pylib/power-twister-harness/README.rst:
Same as Display test harness, this should go into the Twister doc page to be visible.

hakehuang committed:
doc: power_shield: add document to twister

add power_shield documents to twister

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang commented:
@kartben , please help to review. I update the sample code according to your comments. just to clarify, the shield name is not real  shield, it is just follow the form naming with st_powershield. Thanks again

hakehuang committed:
tests: pm: fix power_residency_time on rw612

Use the native RW612 power-state node labels and fall back to stop0/1/2
when a platform still defines those aliases. Also add SPDX headers to the
new requirements files to satisfy compliance checks.

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang commented:
I split this PR into smaller pieces to make review easier:

- #105807 — `samples: adc: add power measure sample`
- #105808 — `tests: scripts: add general ADC power harness support`
- #105809 — `samples: boards: nxp: enable power harness on mimxrt595_evk system_off`
- #105810 — `tests: pm: enable power harness for power_mgmt_soc on mimxrt1060_evk`
- #105811 — `testsuite: add PM power residency measurement support`
- #105812 — `doc: twister: document general ADC power harness usage`

Each PR is scoped to one area only (sample / harness / board integration / PM tests / docs).

I plan to close this larger PR after follow-up review moves to the split PRs.

hakehuang commented:
> Honestly I think this PR is fine as is.

but it just stops here, I guess it may be because there are too much commits in one.

kartben commented:
> but it just stops here, I guess it may be because there are too much commits in one.

But feedback from months ago still applies. It's really not clear at all why you are duplicating an entire code sample (and its metric ton of overlays), with hardly any changes to its documentation or its code.
I would love to spend more time on reviewing it but things like this keep giving me the impression that this PR is not quite ready still?
